File: rtneuron/python/rtneuron/gui/simulationPlayer.py
# -*- coding: utf-8 -*-
## Copyright (c) 2006-2018, École Polytechnique Fédérale de Lausanne (EPFL) /
##                           Blue Brain Project and
##                          Universidad Politécnica de Madrid (UPM)
##                          Juan Hernando <juan.hernando@epfl.ch>
##
## This file is part of RTNeuron <https://github.com/BlueBrain/RTNeuron>
##
## This library is free software; you can redistribute it and/or modify it under
## the terms of the GNU General Public License version 3.0 as published
## by the Free Software Foundation.
##
## This library is distributed in the hope that it will be useful, but WITHOUT
## ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
## FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
## details.
##
## You should have received a copy of the GNU General Public License along
## with this library; if not, write to the Free Software Foundation, Inc.,
## 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

import os
from PyQt5 import QtCore, QtWidgets

import rtneuron as _rtneuron

from .QMLComponent import QMLComponent

class SimulationPlayer(QMLComponent):

    simulation_time_changed = QtCore.pyqtSignal(float, float, float)
    simulation_speed_changed = QtCore.pyqtSignal(float)
    open_simulation_clicked = QtCore.pyqtSignal()

    # internal
    _reset_timestamp = QtCore.pyqtSignal()

    def __init__(self, qml=None, parent=None, view=None, simulation=None):
        super(SimulationPlayer, self).__init__(
                parent, 'SimulatorPlayer.qml' if qml == None else qml)
        self._playing = False
        self._looping = False
        self._engine_player = None

        self.simulation_time_changed.connect(self.qml.onSimulationTimeChanged)
        self.simulation_speed_changed.connect(self.qml.onPlaybackSpeedChanged)
        self.qml.openSimulationClicked.connect(self.open_simulation_clicked)
        self._view = view

        self._reset_timestamp.connect(self._on_reset_timestamp)

    def enable_playback_controls(self, enable):
        if self._view and self._view.scene:
            self._view.attributes.display_simulation = enable

        self.qml.enablePlaybackControls(enable)

    @property
    def view(self):
        return self._view

    @view.setter
    def view(self, view):
        self._view = view
        self.qml.enableOpenButton()

    @property
    def engine_player(self):
        return self._engine_player

    @engine_player.setter
    def engine_player(self, player):
        self._engine_player = player

        """ Connect GUI signals to player actions and engine signals to GUI
        updates.
        Connects the start/stop button click signal and the slider changed
        signal from the GUI and the timestampChanged signal from the player """

        # Connecting QML signals
        def on_play_pause_clicked():
            if self._playing:
                player.pause()
            else:
                player.play()
            self._playing = not self._playing

        def on_loop_button_toggled(checked):
            self._looping = checked

        def on_time_changed(value):
            if player.endTime != player.endTime:
                # Simulation window still not adjusted
                try:
                    player.adjustWindow()
                except RuntimeError:
                    return
            value = min(1, max(value, 0))
            player.timestamp = \
                (player.endTime - player.beginTime) * value + player.beginTime

        def on_speed_changed(val):
            player.simulationDelta = val

        def on_speed_dialog_requested(val):
            simulation_delta, ok = QtWidgets.QInputDialog.getDouble(
                None, 'Simulation Parameters',
                'Simulation delta (ms):', value=val)

            if simulation_delta > 0.0 and ok:
                player.simulationDelta = simulation_delta

        # Connecting slots
        self.qml.playPauseClicked.connect(on_play_pause_clicked)
        self.qml.loopButtonToggled.connect(on_loop_button_toggled)
        self.qml.timeSliderChanged.connect(on_time_changed)
        self.qml.speedInputDialogRequested.connect(on_speed_dialog_requested)
        self.qml.speedSliderChanged.connect(on_speed_changed)

        # Connecting player signals
        def on_timestamp_changed(value):
            begin = player.beginTime
            end = player.endTime
            self.simulation_time_changed.emit(value, begin, end)

        def on_playback_state_changed(state):
            if state == _rtneuron.SimulationPlayer.PLAYING:
                self.qml.setPlaybackState(True)
            elif state == _rtneuron.SimulationPlayer.PAUSED:
                self.qml.setPlaybackState(False)
            elif self._looping and state == _rtneuron.SimulationPlayer.FINISHED:
                # We don't reset from here because setting the timestamp causes
                # a state transition that triggers a Qt warning.
                self._reset_timestamp.emit()

        def on_simulation_delta_value_changed(value):
            self.simulation_speed_changed.emit(value)

        def on_simulation_window_changed(begin, end):
            self.simulation_time_changed.emit(player.timestamp, begin, end)

        player.timestampChanged.connect(on_timestamp_changed)
        player.simulationDeltaChanged.connect(on_simulation_delta_value_changed)
        player.windowChanged.connect(on_simulation_window_changed)
        player.playbackStateChanged.connect(on_playback_state_changed)

        # Try adjusting the simulation window. This will throw if there's
        # no report attached
        if player.endTime != player.endTime:
            try:
                player.adjustWindow()
                # If this succeeds we need to enable the controls
                self.qml.enablePlaybackControls(True)
            except RuntimeError:
                pass

        self.simulation_speed_changed.emit(player.simulationDelta)


    def _on_reset_timestamp(self):
        player = self._engine_player
        player.timestamp = player.window[player.simulationDelta < 0]

File: rtneuron/python/binding/SimulationPlayer.cpp
/* Copyright (c) 2006-2018, Ecole Polytechnique Federale de Lausanne (EPFL) /
 *                           Blue Brain Project and
 *                          Universidad Politecnica de Madrid (UPM)
 *                          Juan Hernando <juan.hernando@epfl.ch>
 *
 * This file is part of RTNeuron <https://github.com/BlueBrain/RTNeuron>
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License version 3.0 as published
 * by the Free Software Foundation.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 * details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this library; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#include "boost_signal_connect_wrapper.h"
#include "docstrings.h"
#include "helpers.h"

#include "rtneuron/SimulationPlayer.h"

#include <boost/python.hpp>
#include <boost/python/enum.hpp>

using namespace boost::python;
using namespace bbp::rtneuron;

// export_SimulationPlayer -----------------------------------------------------

void SimulationPlayer_setWindow(SimulationPlayer* player, object o)
{
    double start, stop;
    extract_pair(o, start, stop);

    /* GIL released to avoid a trivial deadlock if a callback is connected
       to the signal for window changes. */
    ReleaseGIL release;
    player->setWindow(start, stop);
}

tuple SimulationPlayer_getWindow(SimulationPlayer* player)
{
    double start, stop;
    player->getWindow(start, stop);
    return make_tuple(start, stop);
}

void SimulationPlayer_adjustWindow(SimulationPlayer* player)
{
    ReleaseGIL release;
    player->adjustWindow();
}

void SimulationPlayer_play(SimulationPlayer* player)
{
    /* GIL released to avoid a trivial deadlock if a callback is connected
       to the signal for window changes. */
    ReleaseGIL release;
    player->play();
}

void export_SimulationPlayer()
// clang-format off
{

enum_<SimulationPlayer::PlaybackState> playbackState(
    "PlaybackState",
    DOXY_ENUM(bbp::rtneuron::SimulationPlayer::PlaybackState));
playbackState.value("PLAYING", SimulationPlayer::PLAYING);
playbackState.value("PAUSED", SimulationPlayer::PAUSED);
playbackState.value("FINISHED", SimulationPlayer::FINISHED);

class_<SimulationPlayer, SimulationPlayerPtr, boost::noncopyable>
simulationPlayerWrapper(
    "SimulationPlayer", DOXY_CLASS(bbp::rtneuron::SimulationPlayer), no_init);

scope simulationPlayerScope = simulationPlayerWrapper;

scope().attr("PLAYING") = SimulationPlayer::PLAYING;
scope().attr("PAUSED") = SimulationPlayer::PAUSED;
scope().attr("FINISHED") = SimulationPlayer::FINISHED;

WRAP_MEMBER_SIGNAL(SimulationPlayer, PlaybackFinishedSignal)
WRAP_MEMBER_SIGNAL(SimulationPlayer, PlaybackStateChangedSignal)
WRAP_MEMBER_SIGNAL(SimulationPlayer, SimulationDeltaChangedSignal)
WRAP_MEMBER_SIGNAL(SimulationPlayer, WindowChangedSignal)
WRAP_MEMBER_SIGNAL(SimulationPlayer, TimestampChangedSignal)

simulationPlayerWrapper
    .add_property("timestamp",
        &SimulationPlayer::getTimestamp, &SimulationPlayer::setTimestamp,
        (std::string("Get: ") +
         DOXY_FN(bbp::rtneuron::SimulationPlayer::getTimestamp) +
         std::string("\nSet: ") +
         DOXY_FN(bbp::rtneuron::SimulationPlayer::setTimestamp)).c_str())
    .add_property("window",
        &SimulationPlayer_getWindow, &SimulationPlayer_setWindow,
        DOXY_FN(bbp::rtneuron::SimulationPlayer::setWindow))
    .add_property("beginTime",
        &SimulationPlayer::getBeginTime, &SimulationPlayer::setBeginTime,
        DOXY_FN(bbp::rtneuron::SimulationPlayer::setBeginTime))
    .add_property("endTime",
        &SimulationPlayer::getEndTime, &SimulationPlayer::setEndTime,
        DOXY_FN(bbp::rtneuron::SimulationPlayer::setEndTime))
    .add_property("simulationDelta",
        &SimulationPlayer::getSimulationDelta,
        &SimulationPlayer::setSimulationDelta,
        DOXY_FN(bbp::rtneuron::SimulationPlayer::setSimulationDelta))
    .def("adjustWindow", SimulationPlayer_adjustWindow,
         DOXY_FN(bbp::rtneuron::SimulationPlayer::adjustWindow))
    .def("play", SimulationPlayer_play,
         DOXY_FN(bbp::rtneuron::SimulationPlayer::play))
    .def("pause", &SimulationPlayer::pause,
         DOXY_FN(bbp::rtneuron::SimulationPlayer::pause))
    .def_readonly("finished", &SimulationPlayer::finished,
         DOXY_VAR(bbp::rtneuron::SimulationPlayer::finished))
    .def_readonly("playbackStateChanged", &SimulationPlayer::playbackStateChanged,
         DOXY_VAR(bbp::rtneuron::SimulationPlayer::playbackStateChanged))
    .def_readonly("simulationDeltaChanged", &SimulationPlayer::simulationDeltaChanged,
         DOXY_VAR(bbp::rtneuron::SimulationPlayer::simulationDeltaChanged))
    .def_readonly("timestampChanged", &SimulationPlayer::timestampChanged,
         DOXY_VAR(bbp::rtneuron::SimulationPlayer::timestampChanged))
    .def_readonly("windowChanged", &SimulationPlayer::windowChanged,
         DOXY_VAR(bbp::rtneuron::SimulationPlayer::windowChanged))
;

}
// clang-format on
